examples/models/llama2/aqlm/lut_kernel.cpp

```diff
@@ -0,0 +1,148 @@
+#include "lut_kernel.h"
+
+#include <numeric>
+#include <functional>
+#include <omp.h>
+
+#include <executorch/extension/kernel_util/make_boxed_from_unboxed_functor.h>
+#include <executorch/kernels/optimized/blas/CPUBlas.h>
+
+
+template<typename fp_dtype>
+void quadruple_for(
+    int num_inputs,
+    int num_input_groups, const fp_dtype* __restrict__ lut,
+    int out_features, const uint8_t* __restrict__ b_alt,
+    fp_dtype* __restrict__ output_vec
+)
+{
+    std::memset(output_vec, 0, num_inputs * out_features * sizeof(fp_dtype));
+
+    const int lut_stride = num_input_groups * 2 * 256;
+    const int b_alt_stride = 2 * out_features;
+
+    for (int input = 0; input < num_inputs; ++input) {
+        #pragma omp parallel for num_threads(4)
+        for (int j = 0; j < num_input_groups; ++j) {
+            const fp_dtype* lut_ptr = lut + input * lut_stride + j * 2 * 256;
+            const uint8_t* b_alt_ptr = b_alt + j * b_alt_stride;
+
+            for (int i = 0; i < out_features; ++i) {
+                output_vec[input * out_features + i] += lut_ptr[b_alt_ptr[i * 2]];
+                output_vec[input * out_features + i] += lut_ptr[256 + b_alt_ptr[i * 2 + 1]];
+            }
```

Comment:
I would recommend experimenting with unrolling this loop. With clang, you put `#pragma unroll 4` (or whatever unroll count) on the line before the `for`; Google says (https://gcc.gnu.org/onlinedocs/gcc/Loop-Specific-Pragmas.html) the GCC equivalent would be `#pragma GCC unroll 4`.